packages/react/src/combobox/root/ComboboxRootInternal.tsx

```diff
@@ -0,0 +1,1267 @@
+    selectedValue: selectedValueProp,
+    onSelectedValueChange,
+    defaultInputValue = '',
+    inputValue: inputValueProp,
+    selectionMode = 'none',
+    onItemHighlighted: onItemHighlightedProp,
+    name: nameProp,
+    disabled: disabledProp = false,
+    readOnly = false,
+    required = false,
+    inputRef: inputRefProp,
+    cols = 1,
+    items,
+    filter: filterProp,
+    openOnInputClick = true,
+    autoHighlight = false,
+    itemToStringLabel,
+    itemToStringValue,
+    virtualized = false,
+    fillInputOnItemPress = true,
+    modal = false,
+    limit = -1,
+    autoComplete = 'list',
+    locale,
+  } = props;
+
+  const { clearErrors } = useFormContext();
+  const {
+    setDirty,
+    validityData,
+    validationMode,
+    setControlId,
+    setFilled,
+    name: fieldName,
+    disabled: fieldDisabled,
+  } = useFieldRootContext();
+  const fieldControlValidation = useFieldControlValidation();
+
+  const id = useBaseUiId(idProp);
+
+  const disabled = fieldDisabled || disabledProp;
+  const name = fieldName ?? nameProp;
+  const multiple = selectionMode === 'multiple';
+  const commitValidation = fieldControlValidation.commitValidation;
+
+  useIsoLayoutEffect(() => {
+    setControlId(id);
+    return () => {
+      setControlId(undefined);
+    };
+  }, [id, setControlId]);
+
+  const [selectedValue, setSelectedValueUnwrapped] = useControlled<any>({
+    controlled: selectedValueProp,
+    default: multiple ? (defaultSelectedValue ?? EMPTY_ARRAY) : defaultSelectedValue,
+    name: 'Combobox',
+    state: 'selectedValue',
+  });
+
+  const [queryChangedAfterOpen, setQueryChangedAfterOpen] = React.useState(false);
+
+  const collatorFilter = useCoreFilter({ locale });
+
+  const filter = React.useMemo(() => {
+    if (filterProp === null) {
+      return () => true;
+    }
+    if (filterProp !== undefined) {
+      return filterProp;
+    }
+    if (selectionMode === 'single' && !queryChangedAfterOpen) {
+      return createSingleSelectionCollatorFilter(collatorFilter, itemToStringLabel, selectedValue);
+    }
+    return createCollatorItemFilter(collatorFilter, itemToStringLabel);
+  }, [
+    filterProp,
+    selectionMode,
+    selectedValue,
+    queryChangedAfterOpen,
+    collatorFilter,
+    itemToStringLabel,
+  ]);
+
+  const [inputValue, setInputValueUnwrapped] = useControlled({
+    controlled: inputValueProp,
+    default: defaultInputValue,
+    name: 'Combobox',
+    state: 'value',
+  });
+
+  const [open, setOpenUnwrapped] = useControlled({
+    controlled: props.open,
+    default: props.defaultOpen,
+    name: 'Combobox',
+    state: 'open',
+  });
+
+  const [closeQuery, setCloseQuery] = React.useState<string | null>(null);
+
+  const query =
+    closeQuery ?? (inputValue === '' ? '' : String(inputValue).trim().toLocaleLowerCase());
+  const isGrouped = isGroupedItems(items);
+
+  const flatItems: Value[] = React.useMemo(() => {
+    if (!items) {
+      return [];
```

Comment:
Wouldn't `EMPTY_ARRAY` work here?
```suggestion
      return EMPTY_ARRAY;
```